File: Medium/perfect_squares/perfect_squares.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
    public:
        int numSquares(int n) {
            vector<int> a(n+1, n);
            a[0] = 0;
            for(int i = 1; i <= n; i++){
                for(int j = 1; (j*j) <= i; j++){
                    if(a[i] > a[i-(j*j)] + 1){
                        a[i] = a[i-(j*j)] + 1;
                    }
                }
            }
            return a[n];
    }
};

//Example usage:
int main(){
    Solution solution1;
    int n1 = 12;
    int result1 = solution1.numSquares(n1);
    cout << "For the input " << n1 << ", the output is: " << result1 << endl;

    Solution solution2;
    int n2 = 13;
    int result2 = solution2.numSquares(n2);
    cout << "For the input " << n2 << ", the output is: " << result2 << endl;
    return 0;
};

File: Medium/perfect_squares/perfect_squares.py
from typing import List

def numSquares(n: int) -> int:
    a = [n] * (n + 1)
    a[0] = 0

    for i in range(1, n+1):
        j = 1
        while (j*j) <= i:
            if(a[i] > a[i - (j*j)] + 1):
                a[i] = a[i-(j*j)] + 1
            j += 1
    return a[n]

if __name__ == "__main__":
    print("Testing numSquares function:")
    print("For n = 12, the output is:", numSquares(12)) #Output: 3
    print("For n = 13, the output is:", numSquares(13)) #Output: 2
